lib/Producer.js

```diff
@@ -104,22 +104,35 @@ class Producer {
       options = {};
     }
 
-    if (!this.conns || this.conns.length === 0) {
-      return Promise.reject(new Error('No connections to nsqd')).asCallback(callback);
-    }
-
     if (options.retry && this.strategy === Producer.FAN_OUT) {
       return Promise.reject(new Error('Retry on produce level is not supported in fanout strategy')).asCallback(
         callback
       );
     }
 
+    if (!this.conns || this.conns.length === 0) {
+      // Start to reconnect to all the nsqd nodes if there is no available connection.
+      // Will follow the same retry strategy as specifed from the input parameter.
+      return Promise.resolve(
+        promiseRetry(options.retry === true ? {} : options.retry, (retry, number) => {
+          debug(`Attempt to reconnect to all nsqd nodes with retry number ${number}...`);
+          return this.connect().catch(retry);
```

Comment:
Does the `catch` work here?

Reply:
> Does the `catch` work here?

Yes, and it follows the same promiseRetry style as what the current nsq-strategies does in Producer._produceOnce function when it wants to retry publishing the message to nsqd node.

Reply:
> Does the `catch` work here?

After running a few more tests directly on the connect method (Not the mocked one), it does throw `TypeError: fn is not a function` error when calling the promieRetry on the connect method. I will take a deeper look at this tonight and see if I could fix it.

Reply:
> > Does the `catch` work here?
> 
> After running a few more tests directly on the connect method (Not the mocked one), it does throw `TypeError: fn is not a function` error when calling the promieRetry on the connect method. I will take a deeper look at this tonight and see if I could fix it.

It turns out it was the options in the input parameter of the produce method that was causing the above error. Please kindly see the updated changes in this pr later to find more details.

Reply:
Pr is updated now to address the above fix.